src/Token/Builder.php

```diff
@@ -28,9 +28,12 @@ final class Builder implements BuilderInterface
 
     private Parsing\Encoder $encoder;
 
-    public function __construct(Parsing\Encoder $encoder)
+    private bool $dateClaimsWithMicroseconds;
+
+    public function __construct(Parsing\Encoder $encoder, bool $dateClaimsWithMicroseconds)
```

Comment:
I don't think using a boolean here is the way to go... it surely works but it's a smells like something that doesn't belong here.

What do you think about extracting the process of formatting claims into an interface, change `Builder#getToken()` to receive an instance that implements that interface, and change the configuration object to allow users to customise that?

We'd ship here only the formatter using microseconds IMHO but users are free to provide their own implementation and change any claim they desire prior to the encoding process.

Something like:

```php
interface ClaimFormatter
{
    /**
     * @param array<string, mixed> $claims
     *
     * @return array<string, mixed>
     */
    public function formatClaims(array $claims): array;
}

// Terrible name but I hope you can understand the intent
final class AudienceAndDatesWithMicroseconds implements ClaimFormatterintent
{
    public function formatClaims(array $claims): array
    {
        if (count($claims[RegisteredClaims::AUDIENCE]) === 1) {
            $claims[RegisteredClaims::AUDIENCE] = current($claims[RegisteredClaims::AUDIENCE]);
        }

        foreach (array_intersect(RegisteredClaims::DATE_CLAIMS, array_keys($claims)) as $claim) {
            $claims[$claim] = $this->convertDate($claims[$claim]);
        }

        return $claims;
    }

    /** @return int|string */
    private function convertDate(DateTimeImmutable $date)
    {
        $seconds      = $date->format('U');
        $microseconds = $date->format('u');

        if ((int) $microseconds === 0) {
            return (int) $seconds;
        }

        return $seconds . '.' . $microseconds;
    }
}
```

The usage would then be:

```php
$now   = new DateTimeImmutable();

$config->createBuilder()
    ->issuedBy('http://example.com')
    ->issuedAt($now)
    ->expiresAt($now->modify('+1 hour'))
    ->withClaim('uid', 1)
    ->getToken($config->getSigner(), $config->getSigningKey(), $config->getClaimFormatter());
```

What I also like about this design is that it makes it simpler for people to issue tokens with different formatters without changing their configuration.

Reply:
I really like this implementation, and how you extracted the process of formatting claims into an interface. I agree about giving only one implementation of `ClaimFormatter` with this library, at least for now.

Arguments of `Builder#getToken` could be member of the `Builder` class. Maybe we could think about it for the future.

Just one more thing about the `formatClaims` method : from the [RFC](https://tools.ietf.org/html/rfc7519#section-4.1.3), the `aud` claim is optionnal. We should use `isset` function before `count`.